File: app-examples/example-jv-03adapters/src/main/java/foo/bar/example/foreadapters/feature/playlist/MutablePlaylistModel.java
package foo.bar.example.foreadapters.feature.playlist;


import java.util.ArrayList;
import java.util.List;

import co.early.fore.adapters.Adaptable;
import co.early.fore.adapters.mutable.ChangeAwareArrayList;
import co.early.fore.adapters.mutable.ChangeAwareList;
import co.early.fore.adapters.mutable.UpdateSpec;
import co.early.fore.adapters.mutable.Updateable;
import co.early.fore.core.Affirm;
import co.early.fore.core.WorkMode;
import co.early.fore.core.logging.Logger;
import co.early.fore.core.observer.ObservableImp;
import co.early.fore.core.time.SystemTimeWrapper;

import static foo.bar.example.foreadapters.feature.playlist.RandomTrackGeneratorUtil.generateRandomColourResource;

/**
 * Example model based on **mutable** list data
 *
 * Copyright © 2015-2021 early.co. All rights reserved.
 */
public class MutablePlaylistModel extends ObservableImp implements  Adaptable<Track>, Updateable {

    private static final String LOG_TAG = MutablePlaylistModel.class.getSimpleName();

    private final Logger logger;
    private final ChangeAwareList<Track> trackList;


    public MutablePlaylistModel(SystemTimeWrapper systemTimeWrapper, WorkMode workMode, Logger logger) {
        super(workMode, logger);
        this.logger = Affirm.notNull(logger);
        trackList = new ChangeAwareArrayList<>(Affirm.notNull(systemTimeWrapper));
    }

    public void addNewTrack() {
        logger.i(LOG_TAG, "addNewTrack()");
        trackList.add(new Track(generateRandomColourResource()));
        notifyObservers();
    }

    public void removeTrack(int index) {
        logger.i(LOG_TAG, "removeTrack() " + index);
        checkIndex(index);
        trackList.remove(index);
        notifyObservers();
    }

    public void removeAllTracks() {
        logger.i(LOG_TAG, "removeAllTracks()");
        trackList.clear();
        notifyObservers();
    }

    public void increasePlaysForTrack(int index) {
        logger.i(LOG_TAG, "increasePlaysForTrack() " + index);
        getItem(index).increasePlaysRequested();
        trackList.makeAwareOfDataChange(index);
        notifyObservers();
    }

    public void decreasePlaysForTrack(int index) {
        logger.i(LOG_TAG, "decreasePlaysForTrack() " + index);
        getItem(index).decreasePlaysRequested();
        trackList.makeAwareOfDataChange(index);
        notifyObservers();
    }

    public void add5NewTracks() {
        logger.i(LOG_TAG, "add5NewTracks()");
        List<Track> newTracks = new ArrayList<>();
        for (int ii = 0; ii < 5; ii++) {
            newTracks.add(new Track(generateRandomColourResource()));
        }
        trackList.addAll(0, newTracks);
        notifyObservers();
    }

    public void remove5Tracks() {
        logger.i(LOG_TAG, "remove5Tracks()");
        if (getItemCount() > 4) {
            trackList.removeRange(0, 5);
            notifyObservers();
        }
    }

    private void checkIndex(int index) {
        if (trackList.size() == 0) {
            throw new IndexOutOfBoundsException("tracklist has no items in it, can not get index:" + index);
        } else if (index < 0 || index > trackList.size() - 1) {
            throw new IndexOutOfBoundsException("tracklist index needs to be between 0 and " + (trackList.size() - 1) + " not:" + index);
        }
    }

    @Override
    public Track getItem(int index) {
        checkIndex(index);
        return trackList.get(index);
    }

    @Override
    public int getItemCount() {
        return trackList.size();
    }

    @Override
    public UpdateSpec getAndClearLatestUpdateSpec(long maxAgeMs) {
        return trackList.getAndClearLatestUpdateSpec(maxAgeMs);
    }
}

File: app-examples/example-kt-03adapters/src/main/java/foo/bar/example/foreadapterskt/feature/playlist/mutable/MutablePlaylistModel.kt
package foo.bar.example.foreadapterskt.feature.playlist.mutable

import co.early.fore.adapters.Adaptable
import co.early.fore.adapters.mutable.ChangeAwareList
import co.early.fore.adapters.mutable.Updateable
import co.early.fore.kt.core.logging.Logger
import co.early.fore.core.observer.Observable
import co.early.fore.kt.adapters.mutable.ChangeAwareArrayList
import co.early.fore.kt.core.observer.ObservableImp
import foo.bar.example.foreadapterskt.feature.playlist.RandomStuffGeneratorUtil.generateRandomColourResource
import foo.bar.example.foreadapterskt.feature.playlist.RandomStuffGeneratorUtil.randomLong
import foo.bar.example.foreadapterskt.feature.playlist.Track
import java.util.ArrayList

/**
 * Example model based on **mutable** list data
 *
 * Copyright © 2015-2021 early.co. All rights reserved.
 */
class MutablePlaylistModel(
        private val logger: Logger,
        private val trackList: ChangeAwareList<Track> = ChangeAwareArrayList()
) : Observable by ObservableImp(),
    Updateable by trackList,
    Adaptable<Track> {

    fun removeTrack(index: Int) {
        logger.i("removeTrack() $index")
        checkIndex(index)
        trackList.removeAt(index)
        notifyObservers()
    }

    fun removeAllTracks() {
        logger.i("removeAllTracks()")
        trackList.clear()
        notifyObservers()
    }

    fun increasePlaysForTrack(index: Int) {
        logger.i("increasePlaysForTrack() $index")
        getItem(index).increasePlaysRequested()
        trackList.makeAwareOfDataChange(index)
        notifyObservers()
    }

    fun decreasePlaysForTrack(index: Int) {
        logger.i("decreasePlaysForTrack() $index")
        getItem(index).decreasePlaysRequested()
        trackList.makeAwareOfDataChange(index)
        notifyObservers()
    }

    fun addNTracks(n: Int) {
        logger.i("addNTracks() n:$n")
        val newTracks = ArrayList<Track>()
        for (ii in 0 until n) {
            newTracks.add(Track(generateRandomColourResource(), randomLong()))
        }
        trackList.addAll(newTracks)
        logger.i("addNTracks() updated")
        notifyObservers()
    }

    fun removeNTracks(n: Int) {
        logger.i("removeNTracks() n:$n")
        if (itemCount > n - 1) {
            trackList.removeRange(0, n)
            notifyObservers()
        }
    }

    fun isEmpty(): Boolean {
        return trackList.isEmpty()
    }

    fun hasAtLeastNItems(n: Int): Boolean {
        return trackList.size >= n
    }

    override fun getItem(index: Int): Track {
        checkIndex(index)
        return trackList[index]
    }

    override fun getItemCount(): Int {
        return trackList.size
    }

    private fun checkIndex(index: Int) {
        if (trackList.size == 0) {
            throw IndexOutOfBoundsException("tracklist has no items in it, can not get index:$index")
        } else if (index < 0 || index > trackList.size - 1) {
            throw IndexOutOfBoundsException("tracklist index needs to be between 0 and " + (trackList.size - 1) + " not:" + index)
        }
    }
}
